Improve meson-lookup-doc

Janfel opened the pull request:
This new regexp matches objects and methods as well as functions. It also doesn’t match the empty string any more.

This is the original `rx` form that I have converted to `concat`:

```elisp
(rx bol (or (seq "- `" (literal str) "(" )
            (seq (+ ?#) ?  (or (seq (literal str) "()")
                               (seq ?` (literal str) ?`)))))
```

Janfel committed:
Improve meson-lookup-doc regexp

This new regexp matches objects and methods as well as functions.

This is the original `rx` form that I have converted to `concat`:

```elisp
(rx bol (or (seq "- `" (literal str) "(" )
              (seq (+ ?#) ?  (or (seq (literal str) "()")
                                 (seq ?` (literal str) ?`)))))
```

wentasah commented:
Hi!

Thanks for the PR and sorry for delayed response (I had my vacations last weeks).

This is definitely a useful extension. I wonder why you converted the regexp from `rx` to `concat`. The `rx` version is definitely more understandable. You seem to use `literal`, which is only documented in Emacs 27. I'd prefer staying compatible with Emacs 26. I think, you can use `str` without `literal` and `regexp-quote` to stay compatible and, at the same time, end up with a simpler expression.

Additionally, this change would be more useful if you also update `meson-lookup-doc-at-point`. Are you interested in extending the PR with this change?

Janfel commented:
Yes, I would like to extend this PR. I’ve been experimenting with `meson-lookup-doc` for a while now in my personal config.

wentasah commented:
Good - waiting for updates.

Janfel committed:
Introduce meson-lookup-doc helper functions.

Janfel committed:
Rework meson-lookup-doc.

Janfel committed:
Make meson-lookup-doc interactive.

Janfel committed:
Make lookup-doc-at-point an alias to lookup-doc.

Janfel committed:
Update README.md.

wentasah commented on meson-mode.el:
```diff
@@ -892,6 +892,28 @@ Optional SYN-PPSS is the value returned by `syntax-ppss'."
+  "Make regexp for looking up IDENTIFIER in the Meson reference manual."
+  (rx bol (or (+ ?#) ?-) ?  (? ?`)
+      ;; This would be (literal identifier) in Emacs 27.1+.
+      (regexp (regexp-quote identifier))
```
Unfortunately, even this is not supported in Emacs 26. There, `regexp` requires string argument and not arbitrary lisp expression as in v27. I tried using `eval` here, but it evaluates the expression at macro expansion time and complains about `identifier` being void. Currently, I don't see a better way than something like `(concat (rx ...) (regexp-quote indetifier) (rx ...))` or using `rx-to-string`, which at runtime. `rx-to-string` would be OK, because it is not performance critical code.

Now, I'm testing this with:
```elisp
(rx-to-string `(seq bol (or (+ ?#) ?-) ?  (? ?`)
      ;; This would be (literal identifier) in Emacs 27.1+.
      (regexp ,(regexp-quote identifier))
      (or ?\( ?` eol)))
```

Janfel replied:
It seems, that `rx` already calls `regexp-quote` on any string literals. I’ve changed my code to this:
```elisp
(rx-to-string
  `(seq bol (or (+ ?#) ?-) ?  (? ?`) ,identifier (or ?\( ?` eol)))
```

wentasah commented on meson-mode.el:
```diff
@@ -914,21 +914,25 @@ Return either `line-beginning-position' of the matching line or nil."
+  (when-let* ((refman (meson--find-reference-manual))
+	      (buf (find-file-noselect refman))
+	      (pos (with-current-buffer buf
+		     (meson--search-in-reference-manual identifier))))
```
It would be nice to have a `if-let*` here (maybe more than one) with else branch(es) containing error `(message)` like "Meson reference manual not found" and "%s not found in Meson reference manual".

Janfel replied:
I’ve changed the `when-let*` to a `let*` signaling `user-error` in 5f5c1292186ed226ed5ef27e55c2622c80022102.

wentasah commented on meson-mode.el:
```diff
@@ -935,19 +935,13 @@ Return the buffer containing the reference manual or nil."
-  "Show Meson documentation related to current point.
-Currently, it shows something only for functions."
-  (interactive)
-  (if-let ((func (meson-function-at-point)))
```
This removes some functionality that I was using. In the current version, if you type for example: `executable(` or `executable('xxx',` and press F1, the documentation for `executable` function will be shown. With your version, one has to move back to the executable and press F1 there. Can't you do something like calling meson-function-at-point if the identifier in meson-lookup-doc is empty?

Janfel replied:
Fixed in 11cd2ba4fa2195de70ad2a3d193d29082980b4e0.

wentasah reviewed (changes requested):
Overall, it looks good. I have few comments inline. Other minor things:
- Please do not end commit summary lines with dot.
- Can you rebase your branch on top of master? You have the first commit two times in the pull request.

Janfel committed:
Recognize function from within argument list

Janfel committed:
Use rx-to-string in make-lookup-regexp

Janfel committed:
Make lookup-doc signal user-error on failure

Janfel committed:
Untabify line for consistency

Janfel committed:
Remove redundant regexp-quote

wentasah commented on meson-mode.el:
```diff
@@ -925,7 +925,7 @@ Return either `line-beginning-position' of the matching line or nil."
   "Open Meson reference manual and find the function or object named IDENTIFIER.
 Return the buffer containing the reference manual or nil."
-  (interactive (list (thing-at-point 'symbol)))
+  (interactive (list (or (meson-function-at-point) (thing-at-point 'symbol))))
```
I'm not sure about the order here. If I'm inside:
```meson
executable(..., cpp_args : cxx.get_supported_arguments(['-Wsuggest-attribute=const']))
```
then pressing F1 at get_supported_arguments gives me documentation for `executable`, which is probably not, what I wanted. Maybe, with the opposite order, "fails" in some situations as well. What do you think would be better?

Janfel replied:
I have initially chosen this order, partly because `thing-at-point` would match *any* symbol at all and partly because of me misunderstanding your second example. After some experimenting, I now agree that the reverse order is better, as it’s behavior is less surprising to someone used to `thing-at-point` based commands, it enables more granular input based on `point` and it doesn’t need to compute `meson-function-at-point` on every invocation.

Janfel replied:
Changed in 9415e589066bbb62f610353f4daa07b6bc60646b.

wentasah reviewed (commented):
And there is one more thing, which I didn't notice before. Looking up documentation for `executable` does not give documentation for the function _executable_, but for function `build_target`, which has `` `executable` `` listed as a possible value for `target_type` argument. Would it be possible to use different regexps for different lookups? If we know that we are looking for a function (e.g. because there is `(` after the identifier), we would not look for `` `.*` `` and vice versa.

Janfel committed:
Fix regexp matching identifier in backquoted list

Janfel commented:
Good catch. I believe this is a result from me simplifying my original regex too much. This edge case should be fixed by 32ead9e915dfd29fe857bba9b25af7b88228728d.

Janfel committed:
Reverse order of thing-at-point checks

wentasah commented:
Good. Now I'm mostly happy about the functionality. Two minor points remain:
- If I open documentation from a frame with a single window and close it with `q`, then I end up with two windows and with the point in the new window. By repeating this, new windows are being created. I'd prefer that after pressing `q`, the window configuration goes back to the state before opening the documentation.
- The `rx` syntax for characters is not very straightforward. For example in `␣?␣␣` and ``(? ?`)`` the same characters (space (␣) in the former and question mark in the latter) mean different things. I'd prefer having signle-character strings there, i.e. `" "` and ``(? "`")`` for the examples above.

Janfel committed:
Use single character strings

Janfel committed:
Use switch-to-buffer instead of pop-to-buffer

Janfel committed:
Return existing buffer object

Janfel committed:
Set up buffer only once after creation

Janfel committed:
Inline superfluous let-bindings

Janfel committed:
Fix markdown-view-mode not being called

Janfel committed:
Update meson-lookup-doc docstring

Janfel committed:
Regenerate README.md

Janfel commented:
I’ve switched from `pop-to-buffer` to `switch-to-buffer` so the manual buffer will open in the current window again. I’ve also changed the buffer initalization code, so you should be able to control this behavior with `display-buffer-alist` and popup systems using it.

wentasah commented on meson-mode.el:
```diff
@@ -900,34 +900,66 @@ arguments."
+  ;; In Emacs 27 this could be simplified to (rx ... (literal identifier) ...).
+  (rx-to-string
+   `(seq bol (or (seq (+ "#") " " (? "`") ,identifier (or "(" "`" eol))
+                 (seq "- `" ,identifier "(")))))
```
I've found that for matching methods of dependency objects, the following change is needed:
```suggestion
                 (seq (* blank) "- `" ,identifier "(")))))
```

wentasah reviewed (changes requested):
Otherwise, it looks good :-)

While playing with the code, I've found that it cannot find the documentation of a few methods, but this problem is better solved elsewhere: https://github.com/mesonbuild/meson/pull/7700.

Janfel committed:
Match methods of dependency objects

Co-authored-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah reviewed (approved):
Good. Thank you very much for this nice feature. I think I'll make a new release soon :)
